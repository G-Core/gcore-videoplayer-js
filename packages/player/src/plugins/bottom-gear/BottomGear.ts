import { UICorePlugin, template, Events as ClapprEvents, $ } from '@clappr/core'
import { trace } from '@gcorevideo/utils'
import assert from 'assert'

import { CLAPPR_VERSION } from '../../build.js'

import pluginHtml from '../../../assets/bottom-gear/bottomgear.ejs'
import '../../../assets/bottom-gear/gear.scss'
import '../../../assets/bottom-gear/gear-sub-menu.scss'
import gearIcon from '../../../assets/icons/new/gear.svg'
import gearHdIcon from '../../../assets/icons/new/gear-hd.svg'
import { ZeptoResult } from '../../types.js'

const VERSION = '2.19.12'

const T = 'plugins.bottom_gear'

/**
 * Events triggered by the plugin
 * @beta
 */
export enum GearEvents {
  /**
   * Subscribe to this event to accurately attach an item to the gear menu
   */
  RENDERED = 'rendered',
}

// TODO disabled if no items added

/**
 * `PLUGIN` that adds a button to extend the media controls UI with extra options.
 * @beta
 * @remarks
 * The plugin renders small gear icon to the right of the media controls.
 * It provides a base for attaching custom settings UI in the gear menu
 *
 * Depends on:
 *
 * - {@link MediaControl}
 *
 * @example
 * You can use bottom gear to add custom settings UI to the gear menu.
 *
 * ```ts
 * import { BottomGear } from '@gcorevideo/player/plugins/bottom-gear';
 *
 * class CustomOptionsPlugin extends UICorePlugin {
 *   // ...
 *
 *   override get events() {
 *     return {
 *       'click #my-button': 'doMyAction',
 *     }
 *   }
 *
 *   private doMyAction() {
 *     // ...
 *   }
 *
 *   override render() {
 *     const bottomGear = this.core.getPlugin('bottom_gear');
 *     if (!bottomGear) {
 *       return this;
 *     }
 *     this.$el.html('<button class="custom-option">Custom option</button>');
 *     // Put rendered element into the gear menu
 *     bottomGear.addItem('custom').html(this.$el)
 *     return this;
 *   }
 *
 *   // alternatively, add an option with a submenu
 *   override render() {
 *     this.$el.html(template(templateHtml)({
 *       // ...
 *     })));
 *     return this;
 *   }
 *
 *   private addGearOption() {
 *     this.core.getPlugin('bottom_gear')
 *       .addItem('custom', this.$el)
 *       .html($('<button class="custom-option">Custom option</button>'))
 *   }
 *
 *   override bindEvents() {
 *     this.listenToOnce(this.core, ClapprEvents.CORE_READY, () => {
 *       const bottomGear = this.core.getPlugin('bottom_gear');
 *       assert(bottomGear, 'bottom_gear plugin is required');
 *       // simple case
 *       this.listenTo(bottomGear, GearEvents.RENDERED, this.render);
 *       // or with a submenu
 *       this.listenTo(bottomGear, GearEvents.RENDERED, this.addGearOption);
 *      });
 *   }
 * }
 * ```
 */
export class BottomGear extends UICorePlugin {
  private isHd = false

  /**
   * @internal
   */
  get name() {
    return 'bottom_gear'
  }

  /**
   * @internal
   */
  get supportedVersion() {
    return { min: CLAPPR_VERSION }
  }

  /**
   * @internal
   */
  static get version() {
    return VERSION
  }

  private static readonly template = template(pluginHtml)

  /**
   * @internal
   */
  override get attributes() {
    return {
      class: 'media-control-gear',
    }
  }

  /**
   * @internal
   */
  override get events() {
    return {
      'click #gear-button': 'toggleGearMenu',
    }
  }

  /**
   * @internal
   */
  override bindEvents() {
    this.listenToOnce(this.core, ClapprEvents.CORE_READY, this.onCoreReady)
    this.listenTo(
      this.core,
      ClapprEvents.CORE_ACTIVE_CONTAINER_CHANGED,
      this.onActiveContainerChanged,
    )
  }

  /**
   * Adds a custom option to the gear menu
   * @param name - A unique name of the option
   * @param $subMenu - The submenu to attach to the option
   * @returns The added item placeholder to attach custom markup
   * @remarks
   * When called with $submenu param, a click on the added item will toggle the submenu visibility.
   *
   * When added without submenu, it's responsibility of the caller to handle the click event however needed.
   * @example
   * ```ts
   * class MyPlugin extends UICorePlugin {
   *   override render() {
   *     this.$el.html('<div class="my-awesome-settings">...</div>')
   *     this.core.getPlugin('bottom_gear')
   *       ?.addItem('custom', this.$el)
   *       .html($('<button>Custom settings</button>'))
   *     return this
   *   }
   * }
   * ```
   */
  addItem(name: string, $subMenu?: ZeptoResult): ZeptoResult {
    const $existingItem = this.$el.find(`#gear-options li[data-${name}`)
    if ($existingItem.length) {
      trace(`${T} addItem already exists`, { name })
      return $existingItem
    }
    const $item = $('<li></li>')
      .attr(`data-${name}`, '')
      .appendTo(this.$el.find('#gear-options'))
    if ($subMenu) {
      trace(`${T} addItem adding submenu`, { name })
      $subMenu
        .addClass('gear-sub-menu-wrapper')
        .hide()
        .appendTo(this.$el.find('#gear-options-wrapper'))
      $item.on('click', (e: MouseEvent) => {
        trace(`${T} addItem submenu clicked`, { name })
        e.stopPropagation()
        $subMenu.show()
        this.$el.find('#gear-options').hide()
      })
    }
    return $item
  }

  private onActiveContainerChanged() {
    trace(`${T} onActiveContainerChanged`)
    this.bindContainerEvents()
  }

  private bindContainerEvents() {
    trace(`${T} bindContainerEvents`)
    this.listenTo(
      this.core.activeContainer,
      ClapprEvents.CONTAINER_HIGHDEFINITIONUPDATE,
      this.highDefinitionUpdate,
    )
  }

  private highDefinitionUpdate(isHd: boolean) {
    trace(`${T} highDefinitionUpdate`, { isHd })
    this.isHd = isHd
    this.$el.find('.gear-icon').html(isHd ? gearHdIcon : gearIcon)
  }

  /**
   * @internal
   */
  override render() {
    trace(`${T} render`)
    const mediaControl = this.core.getPlugin('media_control')
    if (!mediaControl) {
      return this // TODO test
    }
    const icon = this.isHd ? gearHdIcon : gearIcon
    this.$el
      .html(BottomGear.template({ icon }))
      .find('#gear-sub-menu-wrapper')
      .hide()

    // TODO make non-clickable when there are no items

    setTimeout(() => {
      this.trigger(GearEvents.RENDERED)
    }, 0)

    return this
  }

  /**
   * Collapses any submenu open back to the gear menu.
   * @remarks
   * Should be called by the UI plugin that added a gear item with a submenu when the latter is closed (e.g., when a "back" button is clicked).
   */
  refresh() {
    this.$el.find('.gear-sub-menu-wrapper').hide()
    this.$el.find('#gear-options').show()
  }

  private toggleGearMenu() {
    this.$el.find('#gear-options-wrapper').toggle()
  }

  private hide() {
    this.$el.find('#gear-options-wrapper').hide()
  }

  private onCoreReady() {
    trace(`${T} onCoreReady`)
    const mediaControl = this.core.getPlugin('media_control')
    assert(mediaControl, 'media_control plugin is required')
    this.listenTo(
      mediaControl,
      ClapprEvents.MEDIACONTROL_RENDERED,
      this.onMediaControlRendered,
    )
    this.listenTo(mediaControl, ClapprEvents.MEDIACONTROL_HIDE, this.hide)
  }

  private onMediaControlRendered() {
    trace(`${T} onMediaControlRendered`)
    const mediaControl = this.core.getPlugin('media_control')
    mediaControl.mount('gear', this.$el)
  }
}
