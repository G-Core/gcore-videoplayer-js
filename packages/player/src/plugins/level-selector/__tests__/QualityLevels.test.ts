import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest'
import { Events } from '@clappr/core'
import { QualityLevels } from '../QualityLevels.js'
import {
  createMockBottomGear,
  createMockCore,
  createMockMediaControl,
} from '../../../testUtils.js'
import { GearEvents } from '../../bottom-gear/BottomGear.js'
// import { Logger, LogTracer, setTracer } from '@gcorevideo/utils'

// setTracer(new LogTracer('LevelSelector.test'))
// Logger.enable('*')

const LEVELS = [
  {
    level: 0,
    height: 360,
    width: 640,
    bitrate: 300000,
  },
  {
    level: 1,
    height: 720,
    width: 1280,
    bitrate: 150000,
  },
  {
    level: 2,
    height: 1080,
    width: 1920,
    bitrate: 250000,
  },
]

describe('QualityLevels', () => {
  let core: any
  let levelSelector: QualityLevels
  let mediaControl: any
  let bottomGear: any
  describe('basically', () => {
    beforeEach(() => {
      core = createMockCore({
        levelSelector: {
          // restrictResolution: 360,
          labels: { 720: 'HD', 1080: 'Full HD' },
        },
      })
      core.getPlugin.mockImplementation((name: string) => {
        if (name === 'media_control') {
          return mediaControl
        }
        if (name === 'bottom_gear') {
          return bottomGear
        }
        return null
      })
      mediaControl = createMockMediaControl(core)
      bottomGear = createMockBottomGear(core)
      levelSelector = new QualityLevels(core)
    })
    describe('initially', () => {
      beforeEach(() => {
        core.emit(Events.CORE_READY)
        core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED)
        bottomGear.trigger(GearEvents.RENDERED)
        core.activePlayback.emit(Events.PLAYBACK_LEVELS_AVAILABLE, LEVELS)
      })
      it('should render proper level label', () => {
        expect(
          bottomGear.$el.find('[data-quality]').text(),
          // @ts-ignore
        ).toMatchQualityLevelLabel('auto')
      })
    })
    describe.each([
      ['auto', LEVELS, -1, 'auto'],
      ['standard label', LEVELS, 0, '360p'],
      ['custom label', LEVELS, 1, 'HD'],
    ])('%s', (_, levels, current, label) => {
      beforeEach(() => {
        core.emit(Events.CORE_READY)
        core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED)
        bottomGear.trigger(GearEvents.RENDERED)
        core.activePlayback.emit(Events.PLAYBACK_LEVELS_AVAILABLE, levels)
        levelSelector.$el
          .find(`#level-selector-menu [data-id="${current}"]`)
          .click()
      })
      it('should render the proper level label', () => {
        expect(
          bottomGear.$el.find('[data-quality]').text(),
          // @ts-ignore
        ).toMatchQualityLevelLabel(label)
      })
      it('should render the selected level', () => {
        expect(levelSelector.el.innerHTML).toMatchSnapshot()
        expect(
          levelSelector.$el.find('ul.gear-sub-menu .current')[0].textContent,
          // @ts-ignore
        ).toMatchQualityLevelOption(label)
      })
    })
  })
  describe('options.restrictResolution', () => {
    beforeEach(() => {
      core = createMockCore({
        levelSelector: {
          restrictResolution: 360,
          labels: { 360: '360p', 720: '720p', 1080: '1080p' },
        },
      })
      mediaControl = createMockMediaControl(core)
      bottomGear = createMockBottomGear(core)
      core.getPlugin.mockImplementation((name: string) => {
        if (name === 'media_control') {
          return mediaControl
        }
        if (name === 'bottom_gear') {
          return bottomGear
        }
        return null
      })
      mediaControl = createMockMediaControl(core)
      bottomGear = createMockBottomGear(core)
      levelSelector = new QualityLevels(core)
      core.emit(Events.CORE_READY)
      core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED, core.activeContainer)
      bottomGear.trigger(GearEvents.RENDERED)
    })
    describe('initially', () => {
      beforeEach(() => {
        core.activePlayback.emit(Events.PLAYBACK_LEVELS_AVAILABLE, LEVELS)
      })
      it('should render the restricted quality level label', () => {
        expect(bottomGear.$el.find('[data-quality]').html()).toMatchSnapshot()
        expect(
          bottomGear.$el.find('[data-quality]').text(),
          // @ts-ignore
        ).toMatchQualityLevelLabel('360p')
        expect(
          levelSelector.$el.find('#level-selector-menu .current').text(),
          // @ts-ignore
        ).toMatchQualityLevelOption('360p')
      })
      describe('when opened', () => {
        beforeEach(() => {
          bottomGear.$el.find('[data-quality]').click()
        })
        it('should render the restricted level items disabled', () => {
          expect(levelSelector.el.innerHTML).toMatchSnapshot()
          const allItems = levelSelector.$el.find('#level-selector-menu li')
          expect(allItems.length).toBe(3)
          const unrestrictedItems = allItems.filter(':not(.disabled)')
          expect(unrestrictedItems.length).toBe(1)
          // @ts-ignore
          expect(unrestrictedItems.text()).toMatchQualityLevelOption('360p')
        })
      })
    })
    describe('given vertical video format levels', () => {
      beforeEach(() => {
        core.activePlayback.emit(Events.PLAYBACK_LEVELS_AVAILABLE, [
          {
            level: 0,
            width: 360,
            height: 640,
            bitrate: 450000,
          },
          {
            level: 1,
            width: 720,
            height: 1280,
            bitrate: 150000,
          },
          {
            level: 2,
            width: 1080,
            height: 1920,
            bitrate: 250000,
          },
        ])
      })
      it('should recognize vertical orientation', () => {
        expect(levelSelector.el.innerHTML).toMatchSnapshot()
        expect(
          levelSelector.$el.find('#level-selector-menu [data-id]:eq(0)').text(),
          // @ts-ignore
        ).toMatchQualityLevelOption('1080p')
        expect(
          levelSelector.$el.find('#level-selector-menu [data-id]:eq(1)').text(),
          // @ts-ignore
        ).toMatchQualityLevelOption('720p')
        expect(
          levelSelector.$el.find('#level-selector-menu [data-id]:eq(2)').text(),
          // @ts-ignore
        ).toMatchQualityLevelOption('360p')
      })
      it('should properly apply the restriction', () => {
        expect(
          levelSelector.$el.find('#level-selector-menu li:not(.disabled)')[0]
            .textContent,
          // @ts-ignore
        ).toMatchQualityLevelOption('360p')
      })
    })
  })
})

expect.extend({
  toMatchQualityLevelLabel(received, expected) {
    const { isNot } = this
    const rendered = received
      .replace(/\/assets\/.*\.svg/g, '')
      .replace(/\s+/g, ' ')
      .trim()
    return {
      pass: rendered.includes(`quality ${expected}`),
      message: () =>
        `Quality label must${
          isNot ? ' not' : ''
        } be ${expected} in "${rendered}"`,
    }
  },
  toMatchQualityLevelOption(received, expected) {
    const { isNot } = this
    const rendered = received
      .replace(/\/assets\/.*\.svg/g, '')
      .replace(/\s+/g, ' ')
      .trim()
    return {
      pass: rendered === expected,
      message: () =>
        `Quality option must${
          isNot ? ' not' : ''
        } be ${expected} in "${rendered}"`,
    }
  },
})
