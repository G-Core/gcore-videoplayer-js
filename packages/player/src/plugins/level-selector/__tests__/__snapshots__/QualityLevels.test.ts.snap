// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QualityLevels > basically > auto > should render the selected level 1`] = `
"<button class="gplayer-lite-btn go-back gcore-skin-text-color" id="level-selector-back-button">
    <span class="arrow-left-icon">/assets/icons/new/arrow-left.svg</span>
    quality
</button>
<ul class="gear-sub-menu quality-levels" id="level-selector-menu" role="menu">
    
        <li class="current">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color gcore-skin-active" data-id="-1" id="level_selector_auto" aria-checked="true" role="menuitemradio">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                auto
            </a>
        </li>
    
    
        <li class="">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-id="2" aria-disabled="false" aria-checked="false" role="menuitemradio" id="level_selector_1080">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                Full HD
            </a>
        </li>
    
        <li class="">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-id="1" aria-disabled="false" aria-checked="false" role="menuitemradio" id="level_selector_720">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                HD
            </a>
        </li>
    
        <li class="">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-id="0" aria-disabled="false" aria-checked="false" role="menuitemradio" id="level_selector_360">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                360p
            </a>
        </li>
    
</ul>
"
`;

exports[`QualityLevels > basically > custom label > should render the selected level 1`] = `
"<button class="gplayer-lite-btn go-back gcore-skin-text-color" id="level-selector-back-button">
    <span class="arrow-left-icon">/assets/icons/new/arrow-left.svg</span>
    quality
</button>
<ul class="gear-sub-menu quality-levels" id="level-selector-menu" role="menu">
    
        <li class="">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-id="-1" id="level_selector_auto" aria-checked="true" role="menuitemradio">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                auto
            </a>
        </li>
    
    
        <li class="">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-id="2" aria-disabled="false" aria-checked="false" role="menuitemradio" id="level_selector_1080">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                Full HD
            </a>
        </li>
    
        <li class="current">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color gcore-skin-active" data-id="1" aria-disabled="false" aria-checked="false" role="menuitemradio" id="level_selector_720">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                HD
            </a>
        </li>
    
        <li class="">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-id="0" aria-disabled="false" aria-checked="false" role="menuitemradio" id="level_selector_360">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                360p
            </a>
        </li>
    
</ul>
"
`;

exports[`QualityLevels > basically > standard label > should render the selected level 1`] = `
"<button class="gplayer-lite-btn go-back gcore-skin-text-color" id="level-selector-back-button">
    <span class="arrow-left-icon">/assets/icons/new/arrow-left.svg</span>
    quality
</button>
<ul class="gear-sub-menu quality-levels" id="level-selector-menu" role="menu">
    
        <li class="">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-id="-1" id="level_selector_auto" aria-checked="true" role="menuitemradio">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                auto
            </a>
        </li>
    
    
        <li class="">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-id="2" aria-disabled="false" aria-checked="false" role="menuitemradio" id="level_selector_1080">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                Full HD
            </a>
        </li>
    
        <li class="">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-id="1" aria-disabled="false" aria-checked="false" role="menuitemradio" id="level_selector_720">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                HD
            </a>
        </li>
    
        <li class="current">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color gcore-skin-active" data-id="0" aria-disabled="false" aria-checked="false" role="menuitemradio" id="level_selector_360">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                360p
            </a>
        </li>
    
</ul>
"
`;

exports[`QualityLevels > options.restrictResolution > given vertical video format levels > should recognize vertical orientation 1`] = `
"<button class="gplayer-lite-btn go-back gcore-skin-text-color" id="level-selector-back-button">
    <span class="arrow-left-icon">/assets/icons/new/arrow-left.svg</span>
    quality
</button>
<ul class="gear-sub-menu quality-levels" id="level-selector-menu" role="menu">
    
    
        <li class=" disabled">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-id="2" aria-disabled="true" aria-checked="false" role="menuitemradio" id="level_selector_1080">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                1080p
            </a>
        </li>
    
        <li class=" disabled">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-id="1" aria-disabled="true" aria-checked="false" role="menuitemradio" id="level_selector_720">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                720p
            </a>
        </li>
    
        <li class=" current">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color gcore-skin-active" data-id="0" aria-disabled="false" aria-checked="true" role="menuitemradio" id="level_selector_360">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                360p
            </a>
        </li>
    
</ul>
"
`;

exports[`QualityLevels > options.restrictResolution > initially > should render the restricted quality level label 1`] = `
"<button class="gplayer-lite-btn gcore-skin-text-color gear-option" aria-haspopup="menu">
    <span class="gear-option_icon hidden">/assets/icons/new/hd.svg</span>
    <span class="gear-option_label">quality</span>
    <span class="gear-option_value">360p</span>
    <span class="gear-option_arrow-right-icon">/assets/icons/new/arrow-right.svg</span>
</button>
"
`;

exports[`QualityLevels > options.restrictResolution > initially > when opened > should render the restricted level items disabled 1`] = `
"<button class="gplayer-lite-btn go-back gcore-skin-text-color" id="level-selector-back-button">
    <span class="arrow-left-icon">/assets/icons/new/arrow-left.svg</span>
    quality
</button>
<ul class="gear-sub-menu quality-levels" id="level-selector-menu" role="menu">
    
    
        <li class=" disabled">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-id="2" aria-disabled="true" aria-checked="false" role="menuitemradio" id="level_selector_1080">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                1080p
            </a>
        </li>
    
        <li class=" disabled">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color" data-id="1" aria-disabled="true" aria-checked="false" role="menuitemradio" id="level_selector_720">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                720p
            </a>
        </li>
    
        <li class=" current">
            <a href="#" class="gear-sub-menu_btn gcore-skin-text-color gcore-skin-active" data-id="0" aria-disabled="false" aria-checked="true" role="menuitemradio" id="level_selector_360">
                <span class="check-icon">/assets/icons/new/check.svg</span>
                360p
            </a>
        </li>
    
</ul>
"
`;
